sycl/doc/extensions/proposed/sycl_ext_oneapi_spirv_queries.asciidoc

```diff
@@ -0,0 +1,371 @@
+link:https://www.khronos.org/spirv/[SPIR-V] features that are supported by a
+SYCL device.
+These queries may be useful for applications that use the
+link:../experimental/sycl_ext_oneapi_kernel_compiler_spirv.asciidoc[sycl_ext_oneapi_kernel_compiler_spirv]
+extension to ensure that a SYCL device supports the features required by a
+SPIR-V binary module.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_SPIRV_QUERIES`
+to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Device aspects
+
+This extension adds the following new device aspects.
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl {
+
+enum class aspect : /* unspecified */ {
+  ...
+  ext_oneapi_spirv,
+  ext_oneapi_spirv_queries
+}
+
+} // namespace sycl
+----
+!====
+
+.*aspect::ext_oneapi_spirv*
+Indicates that the device supports compiling and launching kernels written in
+SPIR-V and queries for supported SPIR-V versions.
+
+.*aspect::ext_oneapi_spirv_queries*
+Indicates that the device supports queries for supported SPIR-V extended
+instruction sets, extensions, and capabilities.
+|====
+
+=== Queries
+
+==== New SPIR-V version type
+
+This extension adds the following type and constant definitions to
+identify a SPIR-V version.
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+struct spirv_version {
+  uint16_t major;
+  uint16_t minor;
+};
+
+inline constexpr spirv_version spirv_1_0 = {1,0};
+inline constexpr spirv_version spirv_1_1 = {1,1};
+inline constexpr spirv_version spirv_1_2 = {1,2};
+inline constexpr spirv_version spirv_1_3 = {1,3};
+inline constexpr spirv_version spirv_1_4 = {1,4};
+inline constexpr spirv_version spirv_1_5 = {1,5};
+inline constexpr spirv_version spirv_1_6 = {1,6};
+
+} // namespace ext::oneapi::experimental
+----
+!====
+
+The meaning of the `major` and `minor` values are defined by the SPIR-V
+specification.
+
+The constant values (e.g. `spirv_1_0`) are shorthands that identify various
+SPIR-V versions.
+|====
+
+==== New device information descriptors
+
+This extension adds the following new information descriptors to
+`device::get_info`.
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+namespace sycl::ext::oneapi::experimental::info::device {
+struct spirv_versions {
+  using return_type = std::set<sycl::ext::oneapi::experimental::spirv_version>;
```

Comment:
@bashbaug also brought this up in another comment.  Should we define these to return `std::vector` instead?  That would be more consistent with all the other `get_info` queries.  Since we also have `device::ext_oneapi_supports_spirv_version`, there's not really any need to return a set here.  The main use case for the `get_info` query is to iterate over all the supported versions (e.g. to print them), and that's just as easy to do with a vector.

Reply:
I switched to a `std::vector` in the latest version.  It's easy enough to switch back if desired, though...

Reply:
Actually, if the performance argument convinces either of you, it might even be worth using `std::unordered_set` instead.  I think that can use a hash, whereas the sorted nature of `std::set` means it needs a binary search.

Reply:
I really don't have a strong opinion here - let me know what you prefer.

I initially thought that there might be a performance or implementation benefit with a `std::vector` because it's a closer match to the driver APIs, but because the driver APIs represent strings as `const char*` vs `std::string` we won't directly be able to build the vectors of strings anyhow (we could for non-string values like capabilities, though).

I'm not sure how much this really matters, though, because applications shouldn't be making these queries in performance-critical code.

Reply:
That's a good point. I hadn't considered the potential overhead of copying a `std::set` or `std::unordered_set`, and I'm not sure how it compares to the overhead of copying a `std::vector`.